src/recompiler.ts

```diff
@@ -0,0 +1,262 @@
+import fs from "fs";
+import path from "path";
+import { createHash } from "crypto";
+import { HardhatRuntimeEnvironment, ProjectPathsConfig, TaskArguments } from "hardhat/types";
+import { starknetCompileAction } from "./task-actions";
+import { getArtifactPath, traverseFiles } from "./utils";
+import { ABI_SUFFIX } from "./constants";
+
+interface ContractData {
+    contentHash: string;
+    outputPath: string;
+    abiPath: string;
+    cairoPath?: string;
+    accountContract?: boolean;
+    disableHintValidation?: boolean;
+}
+
+// Cache file name
+const CACHE_FILE_NAME = "cairo-files-cache.json";
+
+export abstract class Cache {
+    protected cache: Record<string, ContractData> = {};
+    public fsPromises = fs.promises;
+
+    constructor(protected hre: HardhatRuntimeEnvironment) { }
+
+    // Returns the contract data from the cache
+    public async getCache(): Promise<Record<string, ContractData>> {
+        await this.loadCache();
+        return this.cache;
+    }
+
+    // Sets the cache
+    public setCache(cacheData: Record<string, ContractData>): void {
+        this.cache = cacheData;
+    }
+
+    // Returns the cache file path
+    private getCacheFilePath(): string {
+        return path.join(this.hre.config.paths.cache, CACHE_FILE_NAME);
+    }
+
+    // Returns the cache directory path
+    private getCacheDirPath(): string {
+        return path.join(this.hre.config.paths.cache);
+    }
+
+    // Loads the cache from the file
+    public async loadCache(): Promise<void> {
+        const cacheDirpath = this.getCacheDirPath();
+        // Creates cache directory if it doesn't exist
+        if (!fs.existsSync(cacheDirpath)) {
+            fs.mkdirSync(cacheDirpath, { recursive: true });
+        }
+
+        const cacheFilePath = this.getCacheFilePath();
+        if (fs.existsSync(cacheFilePath)) {
+            const cacheBuffer = await this.fsPromises.readFile(cacheFilePath);
+            this.setCache(JSON.parse(cacheBuffer.toString() || "{}"));
+        } else {
+            await fs.promises.writeFile(cacheFilePath, JSON.stringify({}) + "\n");
+            this.setCache({});
+        }
+    }
+
+    // Saves the cache to the file
+    public async saveCache(): Promise<void> {
+        const cacheFilePath = this.getCacheFilePath();
+        await this.fsPromises.writeFile(cacheFilePath, JSON.stringify(this.cache, null, " ") + "\n");
+    }
+}
+
+export class Recompiler extends Cache {
```

Comment:
I think recompiler and cache are different entities. This(update of cache) shall be handled via composition rather than inheritance